### ISSUE TITLE
Increase Required PHP version to 5.6

### ISSUE THREAD

**Issue by danieliser** [PopupMaker/Popup-Maker#721]
Keeping in line with WP core, we are increasing the minimum PHP version required. As such we need to ensure it still fails gracefully.

- [x] Import minimum version activation check. We have this done in Ahoy, User Menus & Content Control via the activation methods and a PHP version check.
- [x] Ensure proper messaging to notify users that the plugin now requires PHP X.X.X
- [x] Update readme required version.
- [x] Add class property to `Popup_Maker` for the min version.

**Comment by fpcorso** [comment]
@danieliser Should we release a minor update previous to this one that adds a small notice (either admin or just in our custom notice area) that tells people the minimum version is changing if they are still using a version less than X.X.X?

I would imagine that someone would be upset if they updated the plugin and suddenly cannot display any popups until they can learn about "this PHP thing" to update it. If we can at least find a way to say we tried to let everyone know ahead of time, that would help smooth the transition a little.

**Comment by danieliser** [comment]
@fpcorso Don't think that scenario is possible currently, could be wrong, but WP Core is already enforcing 5.6 minimum to install latest updates for core. So they wouldn't be able to update core by now and would already be aware this may be occuring.

But if you think its worth the effort we can throw one in to a quick patch version.

**Comment by danieliser** [comment]
Further, the plugin simply wouldn't update, their popups wouldn't break using the old version.

> suddenly cannot display any popups until they can learn about "this PHP thing"

They would try to install the update. WP core would say you can't. Nothing changes from before they tried to update.

**Comment by fpcorso** [comment]
@danieliser I believe that the PHP version check was only added in 5.1 when Site Health was merged into core. So, any person who is still using PM on a site below WP 5.1 could still update. According to the wordpress.org stats, that would be 41% of sites which would be about 164,000 of our active installs could potentially have this issue.

**Comment by danieliser** [comment]
Due to some extensions that need updating, the check for PHP version prior to init of the plugin needs to come after including all the required function files. There are functions that if they don't exists will generate fatal errors in older extensions.

To mitigate this we are not bailing out if they have an older version of PHP until after we load those files currently.

Nothing in those files should be incompatible with older versions of PHP at this time.

Further this will require additional changes going forward to eliminate the errors in identified extensions (Aweber for example).

Once those issues have been worked out we can begin moving the activation PHP version check to before those files get loaded.

Lastly we could look at making a new file that loads just those functions required to prevent error from older versions and bail after that is loaded. This way we bail earlier and prevent fatal errors.

**Comment by danieliser** [comment]
Ok some additional notes:

- [ ] Requires updating Videos extension replacing `PUM_AssetCache::reset_cache();` call at `wp-content/plugins/pum-videos/classes/Upgrades.php:125` with the following.

	if ( function_exists( 'pum_reset_assets' ) ) {
		pum_reset_assets();
	}

Otherwise this is what is happening under the hood.

1. Added new `PUM_Install` class that replaces `PUM_Activator` & `PUM_Deactivator` classes. It contains methods to handle activation, deactivation & uninstall callbacks for network & non-network installs. It also contains all the version requirement checks & notice generation functionality.
2. Newly added activation callback that will short circuit the plugins activation if minimum PHP & WP requirements are not met. This will essentially throw a notice to what is required and deactivate the plugin. This should only be needed on fresh installs to sites running below WP v5.1.
3. Just in case added an additional check in the plugin init to abort in case the plugin is active already on a server with older version of PHP. The main plugin file will of course need to stay clear of code that requires later versions of PHP until the bulk of our users have migrated to latest versions of PHP. It will also show an admin notice to indicate something isn't correct.
4. Some older extensions called on functions that didn't exist when the plugin failed to init. To prevent them from throwing errors all of a sudden there is now a failsafes.php file added that includes logicless functions to replace those that are missing and prevent errors. The only one that we couldn't work out properly was for Popup Videos extension noted above.

Next step in v1.10 we will move the main `Popup_Maker`  class to a new file and only run activation checks via vanilla low requirement PHP code, only after init will namespaces etc be available.

**Comment by fpcorso** [comment]
@danieliser, just to make sure I understand the plan: We moved all activation/deactivation into the new PUM_Install class which also contains a few checks and failsafes now. And, the plugin currently does these checks after required files are included due to extensions, right?

So, currently, a site will still function if using PHP prior to 5.6 then, correct? So, if a site is using PM on PHP 5.4 on WP 5.0, they could still update and it wouldn't break anything and the popups will continue to work, for now? Not advocating for us to keeping support 5.0 or older, just making sure i understand the user flow.

Then, 1.10 or later would move the check prior to required files which will make the popups no longer work if they are below our minimum PHP version, right? Since we added the admin notices, that should be enough warning to not be completely jarring as long as the user updates to 1.9 prior to any major breaking changes.

**Comment by fpcorso** [comment]
@danieliser Also, with the new activation check, that fires only when the plugin is activated and will prevent the plugin from being activated if it doesn't meet the required levels, right?

So, if someone on 5.0 were to update the plugin and then, during testing something else, decides to temporarily deactivate this plugin, will it means they will not be able to reactivate it until after they update either PHP or WP, whichever doesn't meet requirements?

**Comment by danieliser** [comment]
@fpcorso 

1. No required files are not loaded until after the check for PHP version passes.
2. A user with PHP 5.4 and WP 5.0 would not see errors, but would see notice that it requires a newer version of PHP. But the plugin wouldn't init functionality. They would need to downgrade PM or upgrade PHP.
3. See above, older versions of PHP will no longer activate the plugin. It also shouldn't cause issues with extensions due to the failsafe.php file that is loaded in those cases.
4. Correct the new activation check will not let the plugin activate at all if PHP is not high enough. This check coupled with the pre-init check covers new installs and updates.
